File: backend/src/main/java/es/udc/tfgproject/backend/model/services/BusinessService.java
package es.udc.tfgproject.backend.model.services;

import java.math.BigDecimal;
import java.time.LocalTime;
import java.util.List;

import es.udc.tfgproject.backend.model.entities.City;
import es.udc.tfgproject.backend.model.entities.Company;
import es.udc.tfgproject.backend.model.entities.CompanyAddress;
import es.udc.tfgproject.backend.model.entities.CompanyCategory;
import es.udc.tfgproject.backend.model.entities.DiscountTicket.DiscountType;
import es.udc.tfgproject.backend.model.entities.Goal;
import es.udc.tfgproject.backend.model.entities.GoalType;
import es.udc.tfgproject.backend.model.entities.Province;
import es.udc.tfgproject.backend.model.exceptions.InstanceNotFoundException;
import es.udc.tfgproject.backend.model.exceptions.PermissionException;

public interface BusinessService {

	Company addCompany(Long userId, String name, Integer capacity, Boolean reserve, Boolean homeSale,
			Integer reservePercentage, Long companyCategoryId, LocalTime openingTime, LocalTime closingTime,
			LocalTime lunchTime, LocalTime dinerTime) throws InstanceNotFoundException;

	Company modifyCompany(Long userId, Long companyId, String name, Integer capacity, Boolean reserve, Boolean homeSale,
			Integer reservePercentage, Long companyCategoryId, LocalTime openingTime, LocalTime closingTime,
			LocalTime lunchTime, LocalTime dinerTime) throws InstanceNotFoundException, PermissionException;

	Company blockCompany(Long userId, Long companyId) throws InstanceNotFoundException, PermissionException;

	Company unlockCompany(Long userId, Long companyId) throws InstanceNotFoundException, PermissionException;

	void deregister(Long userId, Long companyId) throws InstanceNotFoundException, PermissionException;

	Company findCompany(Long userId) throws InstanceNotFoundException;

	Company findCompanyById(Long userId, Long companyId) throws InstanceNotFoundException;

	List<CompanyCategory> findAllCompanyCategories();

	CompanyCategory addCompanyCategory(Long userId, String name) throws InstanceNotFoundException, PermissionException;

	CompanyCategory modifyCompanyCategory(Long userId, Long companyCategoryId, String name)
			throws InstanceNotFoundException, PermissionException;

	CompanyAddress addCompanyAddress(String street, String cp, Long cityId, Long companyId)
			throws InstanceNotFoundException;

	void deleteCompanyAddress(Long userId, Long addressId) throws InstanceNotFoundException, PermissionException;

	Block<CompanyAddress> findCompanyAddresses(Long userId, Long companyId, int page, int size)
			throws InstanceNotFoundException, PermissionException;

	List<City> findAllCities();

	City addCity(Long userId, Long provinceId, String name) throws InstanceNotFoundException, PermissionException;

	City modifyCity(Long userId, Long cityId, Long provinceId, String name)
			throws InstanceNotFoundException, PermissionException;

	List<Province> findAllProvinces();

	Province addProvince(Long userId, String name) throws InstanceNotFoundException, PermissionException;

	Province modifyProvince(Long userId, Long provinceId, String name)
			throws InstanceNotFoundException, PermissionException;

	Goal addGoal(Long userId, Long companyId, DiscountType discountType, BigDecimal discountCash,
			Integer discountPercentage, Long goalTypeId, int goalQuantity)
			throws InstanceNotFoundException, PermissionException;

	Goal modifyGoal(Long userId, Long companyId, Long goalId, DiscountType discountType, BigDecimal discountCash,
			Integer discountPercentage, Long goalTypeId, int goalQuantity)
			throws InstanceNotFoundException, PermissionException;

	Goal findGoal(Long userId, Long companyId, Long goalId) throws InstanceNotFoundException, PermissionException;

	Goal modifyStateGoal(Long userId, Long companyId, Long goalId, String option)
			throws InstanceNotFoundException, PermissionException;

	Block<Goal> findCompanyGoals(Long userId, Long companyId, int page, int size)
			throws InstanceNotFoundException, PermissionException;

	List<GoalType> findAllGoalTypes();

}

File: frontend/src/backend/businessService.js
import {config, appFetch} from './appFetch';

export const addCompany = (name, capacity, reserve, homeSale, reservePercentage, companyCategoryId,
    openingTime, closingTime, lunchTime, dinerTime, onSuccess, onErrors) =>
    appFetch(`/business/companies`,
        config('POST', {name, capacity, reserve, homeSale, reservePercentage, companyCategoryId,
            openingTime, closingTime, lunchTime, dinerTime}),
        onSuccess, onErrors);

export const modifyCompany = (company, onSuccess, onErrors) =>
    appFetch(`/business/companies/${company.id}`, config('PUT', company),
        onSuccess, onErrors);

export const blockCompany = (companyId, onSuccess) =>
    appFetch(`/business/companies/${companyId}/block`, 
        config('POST'), onSuccess);

export const unlockCompany = (companyId, onSuccess) =>
    appFetch(`/business/companies/${companyId}/unlock`, 
        config('POST'), onSuccess);

export const deregister = (companyId, onSuccess, onErrors) =>
    appFetch(`/business/companies/${companyId}`, 
        config('DELETE'), onSuccess, onErrors);

export const findCompany = (onSuccess) => 
    appFetch('/business/companies/company', 
        config('GET'), onSuccess);

export const findCompanyById = (companyId, onSuccess) => 
    appFetch(`/business/companies/${companyId}`, 
        config('GET'), onSuccess);
    
export const findAllCompanyCategories = (onSuccess) => 
    appFetch('/business/companies/categories', 
        config('GET'), onSuccess);

export const addCompanyCategory = (name, onSuccess, onErrors) =>
    appFetch(`/business/companyCategories`,
        config('POST', {name}), onSuccess, onErrors);

export const modifyCompanyCategory = (companyCategory, onSuccess, onErrors) =>
    appFetch(`/business/companyCategories/${companyCategory.id}`, config('PUT', companyCategory),
        onSuccess, onErrors);

export const addCompanyAddress = (street, cp, cityId, companyId, onSuccess, onErrors) =>
    appFetch(`/business/companyAddresses`,
        config('POST', {street, cp, cityId, companyId}),
        onSuccess, onErrors);

export const deleteCompanyAddress = (addressId, onSuccess, onErrors) =>
    appFetch(`/business/companyAddresses/${addressId}`,
        config('DELETE'), onSuccess, onErrors);

export const findCompanyAddresses = (companyId, {page}, onSuccess) => 
    appFetch(`/business/companyAddresses?companyId=${companyId}&page=${page}`, 
        config('GET'), onSuccess);

export const findAllCities = (onSuccess) => 
    appFetch('/business/cities', 
        config('GET'), onSuccess);

export const addCity = (name, provinceId, onSuccess, onErrors) =>
    appFetch(`/business/cities`,
        config('POST', {name, provinceId}), onSuccess, onErrors);

export const modifyCity = (city, onSuccess, onErrors) =>
    appFetch(`/business/cities/${city.id}`, config('PUT', city),
        onSuccess, onErrors);

export const findAllProvinces = (onSuccess) => 
    appFetch('/business/provinces', 
        config('GET'), onSuccess);

export const addProvince = (name, onSuccess, onErrors) =>
    appFetch(`/business/provinces`,
        config('POST', {name}), onSuccess, onErrors);

export const modifyProvince = (province, onSuccess, onErrors) =>
    appFetch(`/business/provinces/${province.id}`, config('PUT', province),
        onSuccess, onErrors);

export const findCompanyGoals = (companyId, {page}, onSuccess) => 
    appFetch(`/business/companyGoals?companyId=${companyId}&page=${page}`, 
        config('GET'), onSuccess);

export const addGoal = (companyId, discountType, discountCash, discountPercentage, goalTypeId, goalQuantity, 
    onSuccess, onErrors) =>{
        appFetch(`/business/goals`,
        config('POST', {companyId, discountType, discountCash, discountPercentage, goalTypeId, goalQuantity}),
        onSuccess, onErrors);
    }

export const modifyGoal = (goal, companyId, onSuccess, onErrors) =>
    appFetch(`/business/goals/${goal.id}`, config('PUT', {goal, companyId}),
        onSuccess, onErrors);

export const findGoal = (goalId, companyId, onSuccess) =>
    appFetch(`/business/goals/${goalId}?companyId=${companyId}`, 
        config('GET'), onSuccess);

export const modifyStateGoal = (goalId, companyId, option, onSuccess) =>
    appFetch(`/business/goals/${goalId}/modifyStateGoal`, 
        config('POST', {companyId, option}), onSuccess);

export const findAllGoalTypes = (onSuccess) => 
    appFetch('/business/goals/goalTypes', 
        config('GET'), onSuccess);